### PR TITLE
Add option for exhausting on duplicate candidate.

### PR DESCRIPTION
**Pull request by tarheel**
Also cleaned up some random things I encountered while touching these files.

### CODE REVIEW

**Comment by tarheel** [comment]
@HEdingfield @moldover ping! The actual code changes in this PR are pretty simple; it just looks lengthy because it includes a bunch of really minor formatting changes to the documentation.

File: src/com/rcv/Tabulator.java [hunk 1]
```diff
@@ -715,13 +715,32 @@ private boolean hasContinuingCandidates(SortedMap<Integer, Set<String>> rankToCa
       // this is used to determine how many skipped rankings occurred in the case of
       // undervotes
       int lastRank = 0;
-      // loop over all rankings in this CVR from most preferred to least and see how they will
-      // be rendered
+      // candidatesSeen is the set of candidates we've encountered while processing this CVR
+      // in this round; only relevant if exhaustOnDuplicateCandidate is enabled
+      Set<String> candidatesSeen = new HashSet<>();
+      // loop over all rankings in this CVR from most preferred to least until we reach a decision
```
Comment (HEdingfield) [inline src/com/rcv/Tabulator.java]: Is this comment related to the one below it? The wording is a bit confusing...

File: src/com/rcv/Tabulator.java [hunk 2]
```diff
@@ -715,13 +715,32 @@ private boolean hasContinuingCandidates(SortedMap<Integer, Set<String>> rankToCa
...
+          for (String candidate : candidateIDSet) {
+            if (candidatesSeen.contains(candidate)) {
+              duplicateCandidate = candidate;
+              break; // finding one duplicate is enough
+            }
+            candidatesSeen.add(candidate);
+          }
+          if (duplicateCandidate != null) {
+            cvr.exhaust("duplicate candidate: " + duplicateCandidate);
+            break;
```
Comment (HEdingfield) [inline src/com/rcv/Tabulator.java]: Is the intention here that the entire ballot is thrown out completely and not counted, or that it's counted until it sees a duplicate; which means the first time it see it the vote for the duplicate candidate it counts at that level?

If it's the former (which I think makes the most sense), I think we either need to break twice out of here, or place this whole block above the `if (cvr.isExhausted())` check (or add another one below it).

If it's the latter, is there a guarantee that the for loop would traverse `cvr.rankToCandidateIDs.keySet()` in the desired order?

File: docs/config_file_documentation.txt [hunk 3]
```diff
@@ -64,41 +64,42 @@ Config file must be valid JSON format.  Examples can be found under the test fol
...
-       Example: "ES&S"
+        index of the column that contains ballot top-ranked candidate
+        example: 3
+        value: [0..1000]
 
       "precinctColumnIndex" required if "tabulateByPrecinct" is in use
-       the index of the column that contains ballot precinct name
-       example: 2
-       value: [0..1000]
+        the index of the column that contains ballot precinct name
```
[the remark this answers is not in the document]
Comment (tarheel) [reply docs/config_file_documentation.txt]: > Can we include default values in this doc for all the optional parameters? I see it listed for some, but not all.

OK, I'll put "none" for most of these.

> Also wondering if there's any mechanism in place to make sure we update the default values here in the doc if we modify them?

Hmm. Not sure how we would do that.

File: src/com/rcv/ElectionConfig.java [hunk 4]
```diff
@@ -376,10 +371,11 @@ String getUndervoteLabel() {
   // purpose: getter for treatBlankAsUWI rule
   // returns: return true if we are to treat blank cell as UWI
   boolean isTreatBlankAsUWIEnabled() {
```
[the remark this answers is not in the document]
Comment (tarheel) [reply src/com/rcv/ElectionConfig.java]: My personal preference is to use capital letters for the whole initialism; I find it easier to read. If you have a particular reason to care about this, though, let me know.

File: src/com/rcv/RawElectionConfig.java [hunk 5]
```diff
@@ -39,6 +39,8 @@
   public List<CVRSource> cvrFileSources;
   // should we report round-by-round results by precinct also?
   public boolean tabulateByPrecinct;
+  // shall we treat blank cells as UWIs?
+  public boolean treatBlankAsUndeclaredWriteIn;
```
[the remark this answers is not in the document]
Comment (tarheel) [reply src/com/rcv/RawElectionConfig.java]: Yeah, the parser breaks if they're not, for some reason.

File: src/com/rcv/Tabulator.java (hunk 1 above)
Comment (tarheel) [reply src/com/rcv/Tabulator.java]: I'll combine them so that it's clearer.

File: src/com/rcv/Tabulator.java (hunk 2 above)
Comment (tarheel) [reply src/com/rcv/Tabulator.java]: It's the latter -- it should be treated normally until the tabulator happens to encounter a duplicate. I don't think I've seen any rule variants yet in which the tabulation is supposed to look ahead for errors; as far as I know, you always consider only as much of the ballot as is necessary to reach a decision in each round.

The for loop is guaranteed to traverse `cvr.rankToCandidateIDs.keySet()` in the right order because `rankToCandidateIDs` is a SortedMap.

File: src/com/rcv/ElectionConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (HEdingfield) [reply src/com/rcv/ElectionConfig.java]: I prefer that aesthetically too, but was reprimanded by some serious Java guys at work about it. Here's an SO thread on it: https://stackoverflow.com/questions/15526107/acronyms-in-camelcase

[the remark this answers is not in the document]
Comment (tarheel) [reply src/com/rcv/ElectionConfig.java]: Yeesh. I bet @moldover would love to discuss this at great length.

I guess the "camel case algorithm" argument is reasonable for a large-scale system, though not relevant for us... but if we're going to follow Google's style, OK, fine.

File: src/com/rcv/Tabulator.java (hunk 2 above)
Comment (HEdingfield) [reply src/com/rcv/Tabulator.java]: Ok, so the implicit assumption here is that the first time a duplicate value is encountered is the correct one; thus the bias is towards a higher valuation of the duplicated candidate. I'd argue that the duplicated candidate should be thrown out of the vote entirely (if not the whole vote), but if this is a standard known rule in IRV elections, I guess it is what it is.

File: src/com/rcv/RawElectionConfig.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (HEdingfield) [reply src/com/rcv/RawElectionConfig.java]: That's really bizarre... I wonder if it's just a transient IntelliJ error? The real acid test is if the code compiles and works properly.

[the remark this answers is not in the document]
Comment (tarheel) [reply src/com/rcv/RawElectionConfig.java]: If you make them not public, the code still compiles, but the parser will fail to populate the values (which our code doesn't handle gracefully).

File: src/com/rcv/Tabulator.java (hunk 2 above)
Comment (tarheel) [reply src/com/rcv/Tabulator.java]: Yeah. Perhaps other jurisdictions have other variants, but so far this is all that George has mentioned.

File: src/com/rcv/RawElectionConfig.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (HEdingfield) [reply src/com/rcv/RawElectionConfig.java]: Oh, I see, I thought you meant the IntelliJ parser. Again, that's still strange. Maybe worth filing a very low priority bug about if we all agree this is non-intuitive.

[the remark this answers is not in the document]
Comment (tarheel) [reply src/com/rcv/RawElectionConfig.java]: I think this explains it: http://www.baeldung.com/jackson-field-serializable-deserializable-or-not

One of us can clean this up in a future PR.

#73
